File: 정민선/프로그래머스/[PCCP 기출문제] 2번 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    games = len(diffs)
    
    def playGame(diff, time, level, time_prev):
        game_round = diff-level
        taken = 0
        
        if game_round > 0:
            taken += (time+time_prev) * game_round
        
        taken += time
        return taken
        
    left = 1
    right = max(diffs)
    answer = right
    
    while (left < right):
        mid = (left+right)//2
        total = 0
        
        for game in range(games):
            diff = diffs[game]
            time = times[game]
            time_prev = 0

            if game !=  0:
                time_prev = times[game-1]

            total += playGame(diff, time, mid, time_prev)

        if total <= limit:
            answer = min(mid, answer)
            right = mid
        else:
            left = mid + 1

    return answer

File: 이진영/프로그래머스/[PCCP 기출문제] 2번 - 퍼즐 게임 챌린지.js
function solution(diffs, times, limit) {
    // 숙련도의 최솟값
    let l = 1;
    let r = 100000;
    let min = 100000;
    
    while (l <= r) {
        const mid = Math.floor((l + r) / 2);
        
        const t = getTime(mid, diffs, times);
        
        if (t > limit) {
            l = mid + 1;
        } else {
            r = mid - 1;
            min = mid;
        }
    }
    
    return min;
}

const getTime = (level, diffs, times) => {
    let time = 0;
    for (let i = 0; i < diffs.length; i++) {
        if (level >= diffs[i]) {
            time += times[i];
        } else {
            time += (times[i - 1] + times[i]) * (diffs[i] - level) + times[i];
        }
    }
    return time;
}